java-checks/src/main/java/org/sonar/java/checks/MissingBeanValidationCheck.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2019 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import java.text.MessageFormat;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Stream;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.helpers.SymbolUtils;
+import org.sonar.java.resolve.JavaSymbol;
+import org.sonar.java.resolve.JavaType;
+import org.sonar.java.resolve.ParametrizedTypeJavaType;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.semantic.SymbolMetadata;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.ParameterizedTypeTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.TypeTree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S5128")
```

Comment:
> having to manually set up exclusions

I think you misunderstood my intention, or I did not make it clear enough. Let me try again :-)

The idea is not to set up exclusions (blacklisting), but to limit the rule scope (whitelisting) using SonarQube's generic feature.

Maybe an example makes it clearer. This is how my company could set it up for our main product:

> Our architecture has several modules which are called "webapp" where transfer objects - special POJOs intended for (de)serialization - and resource classes (JAX-RS endpoints) reside. It is only in those classes that Bean Validation is used. If we fear (or experience) that S5128 results in false positives in other modules of the application, we could simply do the following:
> * In the SonarQube project, navigate to Administration, General Settings, Analysis Scope, Restrict Scope of Coding Rules, set **Rule Key Pattern** to `squid:S5128` and **File Path Pattern** to `**/webapp/**/*.java` and click **Save**.

If I understood your remark correctly, this is exactly the approach you suggested, but without adding any complexity to the rule itself.

---

However, to answer your question about adding a rule parameter, I am very skeptical about that approach:

* It would add a lot of complexity to the rule code.
* It may not be flexible enough for some users.
* Using the generic "rule parameter" UI for this could result in a bad user experience and could lead to configuration mistakes.
* But the worst part for me is that adding such logic to a rule could set a bad precedent: users could (rightly) argue that this or that controversial rule should have that feature as well.

Reply:
Well, on this I don't fully agree. And maybe I haven't been clear enough as well, so I'll retry too! :smile: 

> The idea is not to set up exclusions (blacklisting), but to limit the rule scope (whitelisting) using SonarQube's generic feature.

On this I'm fully with you, I definitely prefer a whitelisting rather than a blacklisting. The thing I'm more interested in is the HOW we configure the rule. My point is: the rule is not enabled by default, the main reason being that you really need to carefully configure it to target the correct packages.
If we choose to rely on the SQ native approach, it means that we need to proceed in two steps:
1. Go through the rule list, enable the rule in your Quality Profile, remember the rule key
1. For each SonarQube project using beans, navigate to Administration, General Settings, Analysis Scope, Restrict Scope of Coding Rules, set `Rule Key Pattern` to `squid:S5128` and `File Path Pattern` to `**/webapp/**/*.java` (or whatever path depending of the project) and click Save

While it could be possible to do everything in one go:
1. enable the rule in your Quality Profile, and configure the parameter with the whitelist directly.

I know we transfer some logic to the rule, making it slightly more complex, while it's natively supported by SonarQube. But we also reduce the number of places we need to do adjustments in case of misconfiguration (only the rule, instead of all projects). Another aspect is that you will directly see the configuration of the rule while browsing rules, while the Analysis scope pattern is hidden in the settings of a project.

However, and I only realize it now (sorry), the main difference between the two approaches being that using the project analysis scope, you can re-use the same Quality Profile for all your projects without having to update the rule configuration in the QualityProfile... On this, I truly don't know what is the best option. I would personally go for the second ("_mine_"), but I'm fine going with the first ("_yours_") as well. So if I do not convince you, let it that way and I'll validate it!

> But the worst part for me is that adding such logic to a rule could set a bad precedent: users could (rightly) argue that this or that controversial rule should have that feature as well.

Well, there is a few rules already which are using this, and we don't do that much maintenance on it.

Reply:
Okay, I understand your rationale now, but I'm not convinced that we should add a whitelist to the rule itself*.

My gut feeling is that we are trying to solve a problem that people will probably not really have. In cases like this, I try to avoid speculating and instead to gather more insight before implementing things (along the lines of "YAGNI" and "minimum viable product" principles). Therefore, as the rule is already useful without its own whitelist, let's ship it and wait for feedback.

If it turns out a whitelist (or something else) is indeed needed, it should still be easy to add it. As a bonus, we will then have more insight into real use cases, and thus may end up implementing some other, better solution that the two of us of course cannot envision today.

----

<sup>* In fact, I think your ideas on where to configure whitelists are interesting and should really be pursued on the SonarQube level. Together with the observation that some rules already have them, they should feed into a discussion of potential new features / UX reorg. But that's off-topic here, of course.</sup>

Reply:
Indeed, that's a very good point. I'm going to check the rules we have which implements such approach, and start discussing around it internally.